README
Improvements over the original bc:

 * support hexadecimal numbers: 0xABCD,  or 0xabcd
 * support binary numbers: 0b1010101
 * support octal numbers: 0777
 * support scientific notation:  1.23e-1  or  4.56E10
 * added 'and', 'or' and 'xor' bitwise operators
 * when a hex, binary or octal number is encountered, the output
   will be the decimal number plus the last value encoutered.
 * added a commandline '-e' eval expression option.
 * more windows friendly: support CR in input
 * added 'pi'
 * added 'modinv' and 'gcd' functions.


GNU bc version 1.07:

Extra configuration options:

	--with-readline tells bc to use the readline package that allows
	  for editing input lines when run interactive.

	--with-editline tells bc to use the BSD editline package that
	  allows for editing input lines when run interactive.

Extra make steps:

	The simple make compiles a version of bc with fixed parameters
	for the recursive multiplication algorithm.  The fixed parameter
	is the number of digits where a sequential algorithm is used
	instead of the recursive algorithm.  It is set to a value that
	is known good on a couple of machines. (Sparc Ultra 10, Pentium
	II, 450.)  I'm calling this point the crossover point.

	To make a version of bc with a custom crossover point for your
	machine, do the following steps:

		make timetest
		make

	The timetest step takes a minimum of 10 minutes to complete.


-------- Original comp.sources.reviewed README --------

Program: GNU bc
Author: Philip A. Nelson
E-mail: phil@cs.wwu.edu
OS: UNIX (BSD, System V, MINIX, POSIX)
Copying: GNU GPL version 2
Copyright holder: Free Software Foundation, Inc. 
Version: bc version 1.01
Required: vsprintf and vfprintf routines.
Machines: It has been compiled and run on the following environments:
	BSD4.3 (VAX 11)
	MINIX 1.5 (IBM PC, both K&R and ANSI compilers)
	MINIX 1.5 (pc532)
	SUN-OS 4.1 (SUN 3 and SUN 4)
	SVR3V5 (Motorola 68K)
	SVR3.2 (3B2)
	SVR4.0.2 (a 386 box)
	ULTRIX 4.1 (DEC 5000)
	UTS (Amdahl)

bc is an arbitrary precision numeric processing language.  Syntax is
similar to C, but differs in many substantial areas.  It supports
interactive execution of statements.  bc is a utility included in the
POSIX P1003.2/D11 draft standard.

This version was written to be a POSIX compliant bc processor with
several extensions to the draft standard.  Option flags are available
to cause warning or rejection of the extensions to the POSIX standard.
For those who want only POSIX bc with no extensions, a grammar is
provided for exactly the language described in the POSIX document.
The grammar (sbc.y) comes from the POSIX document.  The Makefile
contains rules to make sbc.  (for Standard BC)

Since the POSIX document does not specify how bc must be implemented,
this version does not use the historical method of having bc be a
compiler for the dc calculator.  This version has a single executable
that both compiles the language and runs the a resulting "byte code".
The "byte code" is NOT the dc language.

Also, included in the initial distribution is the library file
vfprintf.c for MINIX systems.  My minix 1.5 did not have this file.
Also, you should verify that vsprintf.c works correctly on your
system.

The extensions add some features I think are missing.  The major
changes and additions for bc are (a) names are allowed to be full
identifiers ([a-z][a-z0-9_]*), (b) addition of the &&, ||, and !
operators, (c) allowing comparison and boolean operations in any
expression, (d) addition of an else clause to the if statement, (e)
addition of a new standard function "read()" that reads a number from
the standard input under program control, (f) passing of arrays as
parameters by variable, (g) addition of the "halt" statement that is
an executable statement unlike the quit (i.e.  "if (1 == 0) quit" will
halt bc but "if (1 == 0) halt" will not halt bc.), and (h) the
addition of the special variable "last" that is assigned the value of
each print as the number is printed.
-----------------------------------------------------------------------


ChangeLog
Thu Apr 21 13:45:18 2005  Willem Hengeveld <itsme@xs4all.nl>
	* fixed some warnings and errors in bc.y when compiling with bison
	* made to work with readline >= 4.3
	* added bitwise 'and', 'or' and 'xor' operators
	* added support for 0xhexval input, when 0xhexval constants
	  are found, output is printed in both hex, and decimal.
	* added '-e' option, to evaluate expression from commandline
	* does not complain about input with CR's in it.
	* added support for 0bBINARY 0OCTAL.
	* hexadecimal numbers now can be lowercase.
	* signs for hex numbers are printed in the right location.
	* added support for 1.23e5  style numbers.
	* fixed bug with uninitialized 'f_void' flag.
	* added 'pi'
	* added 'modinv' and 'gcd' functions.

Fri Apr  7 15:14:43 PDT 2017  Phil Nelson <philnelson@acm.org>
	* NEWS: update for 1.07.1 and 1.4.1
	* FAQ: add a couple new FAQs.
	* doc/bc.1: change doucumentation of length() and correctly
	  represent newline.
	* doc/bc.texi: change doucumentation of length()


Fri Apr  7 13:43:31 PDT 2017 Ken Pizzini <ken@gnu.org>
	* bc/configure.ac: update dc version to 1.4.1;
	  update dc copyright year list

Fri Apr  7 13:14:41 PDT 2017 Phil Nelson <philnelson@acm.org>
	* bc/Makefie.am: Do a different fix for parallel make issues.

Fri Apr  7 00:11:49 PDT 2017 Ken Pizzini <ken@gnu.org>
	* dc/eval.c, dc/numeric.c: take more care to never pass
	  a sign-extended char to the <ctype.h> isXXX() functions.
	  Thanks to a NetBSD compiler warning for flagging this.

Thu Apr  6 22:47:46 PDT 2017 Phil Nelson <philnelson@acm.org>
	* bc/execute.c: More tweaks to fix for read() problems.
	* bc/Makefile.am: Attempt to fix parallel make problems.

Thu Apr  6 16:39:25 PDT 2017 Phil Nelson <philnelson@acm.org>
	* bc/execute.c: Correct leading space problem caused by last fix.
	* bc/configure.ac: version to 1.07.1

Thu Apr  6 15:37:22 PDT 2017 Phil Nelson <philnelson@acm.org>
	* bc/execute.c: Correct input_char for base 36 input.
	* doc/bc.1: Update documentation for base 36 input.
	* doc/bc.texi: Update documentation for base 36 input.

Thu Apr  6 14:26:27 PDT 2017 Ken Pizzini <ken@gnu.org>
	* dc/eval.c: The 'Q' command "fix" of 2016-09-24 was incorrect.
	  It fixed a reported bug, but introduced a regression for other
	  existing (and correct) code.  Let's try this again (sigh).

Thu Apr  6 13:11:15 PDT 2017 Ken Pizzini <ken@gnu.org>
	* dc/dc.c: bah, nevermind; just re-disovered commit of
	  Oct 26, 2013 where the setvbuf() was rendered unnecessary
	  by introduction of explicit fflush()-es.

Thu Apr  6 13:11:15 PDT 2017 Ken Pizzini <ken@gnu.org>
	* dc/dc.c: re-instate setvbuf() call that got lost

Mon Jan 16 14:30:00 PDT 2017 Phil Nelson <philnelson@acm.org>
	* Updated copyright to 2017.

Sun Oct 23 16:38:05 PDT 2016 Ken Pizzini <ken@gnu.org>
	* dc/eval.c: dc_system() already eats up to (and including)
	  a trailing newline, so don't fall-through in dc_evalstr()'s
	  DC_SYSTEM case to DC_COMMENT's skip_past_eol() call,
	  as that would unexpectedly require eating a second newline.
	  Thanks to David José for noticing and reporting the bug.

Thu Oct 20 03:11:43 PDT 2016 Ken Pizzini <ken@gnu.org>
	* doc/dc.texi: fix some typos.

Mon Oct 17 00:28:50 PDT 2016 2016 Ken Pizzini <ken@gnu.org>
	* doc/dc.1, doc/dc.texi: document the new default-to-zero
	  of the 'l' command.

Mon Oct 17 00:02:11 PDT 2016 Ken Pizzini <ken@gnu.org>
	* doc/dc.1, doc/dc.texi: be more careful in specifying how
	  the precision of the result of the "v" (square root)
	  command is determined.
	  Thanks to David José for pointing out that the old text
	  underspecified this.

Sun Oct 16 23:43:05 PDT 2016 Ken Pizzini <ken@gnu.org>
	* dc/stack.d: return a 0 (zero value; instead of reporting an error)
	  for the "l" command when the named register stack is empty.
	  Thanks to David José for the suggestion.  (This is a useful default
	  value, and make GNU dc more compatable with BSD dc and Bell Labs
	  version 7 dc.)

Wed Sep 28 13:24:50 PDT 2016 Ken Pizzini <ken@gnu.org>
	* dc/eval.c, doc/dc.1, doc/dc.texi: update copyright dates

Sat Sep 24 16:04:01 PDT 2016 Ken Pizzini <ken@gnu.org>
	* dc/dc.c, dc/eval.c: yesterday's fix for 'Q' created a
	  new bug for the 'q' command; this fresh (and simpler!)
	  fix does the right thing.  (Thanks again to Paul Rayner.)

Fri Sep 23 22:49:17 PDT 2016 Ken Pizzini <ken@gnu.org>
	* dc/dc.c: don't prematurely exit from stack-unwinding 'Q'
	  command
	* dc/eval.c: ensure that DC_FAIL API value does not
	  collide with any internal dc_status code
	* doc/dc.1, doc/dc.texi: clarify that 'Z' command reports
	  on digits in decimal representation in number (as opposed
	  to ambiguous, perhaps "o"base, digits).
	* (Bugs reported by Paul Rayner; thanks.)

Mon Feb 29 15:56:00 PST 2016 Phil Nelson <philnelson@acm.org>
	* Makefile.am, bc/Makefile.am: update to not include
	libmath.h in distribution.
	* configure.ac: update copyright with 2016

Mon Feb 29 13:06:00 PST 2016 Phil Nelson <philnelson@acm.org>
	* bc/Makefile: Add $(LIBBC) as dependency to fbc so
	libmath.h gets correctly built.  sbc also needed it.

Mon Feb 29 12:05:00 PST 2016 Phil Nelson <philnelson@acm.org>
	* bc/util.c, bc.y: Fix two memory leaks submitted to
	Gentoo by Bruce Dubbs.

Tue Jan 19 22:10:10 PST 2016 Phil Nelson <philnelson@acm.org>
	* bc/{execute.c, load.c, main.c, proto.h, scan.l, storage.c, util.c},
	  lib/number.c: Update copyrights, remove old style function
	declarations, remove _PROTOTYPE macros, update to GPL V3.
	Various other code cleanups.

Mon Jan 11 14:01:32 PST 2016 Ken Pizzini <ken@gnu.org>
	* doc/dc.texi: fix typo (reported by Avinash Sonawane)

Mon May 11 16:40:31 PDT 2015 Ken Pizzini <ken@gnu.org>
	* dc/stack.c: catch the case of an empty value on top of
	  a register stack (i.e., the stack entry is first created by
	  an array store on a previously unallocated register)
	  as a normal run-time error, rather than an abort() condition.
	  (Bug reported by Hanno Böck, who found it during fuzz testing.)


bc/main.c
/*  This file is part of GNU bc.

    Copyright (C) 1991-1994, 1997, 2006, 2008, 2012-2017 Free Software Foundation, Inc.

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 3 of the License , or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; see the file COPYING.  If not, see
    <http://www.gnu.org/licenses>.

    You may contact the author by:
       e-mail:  philnelson@acm.org
      us-mail:  Philip A. Nelson
                Computer Science Department, 9062
                Western Washington University
                Bellingham, WA 98226-9062
       
*************************************************************************/

/* main.c: The main program for bc.  */

#include "bcdefs.h"
#include <signal.h>
#include <errno.h>
#include "proto.h"
#include "getopt.h"


/* Variables for processing multiple files. */
static char first_file;

/* Points to the last node in the file name list for easy adding. */
static file_node *last = NULL;

#if defined(LIBEDIT)
/* The prompt for libedit. */
char el_pmtchars[] = "";
static char *el_pmtfunc(void);
static char *el_pmtfunc(void) { return el_pmtchars; }
#endif

/* long option support */
static struct option long_options[] =
{
  {"compile",     0, &compile_only, TRUE},
  {"help",        0, 0,             'h'},
  {"interactive", 0, 0,             'i'},
  {"expression",  0, 0,             'e'},
  {"mathlib",     0, &use_math,     TRUE},
  {"quiet",       0, &quiet,        TRUE},
  {"standard",    0, &std_only,     TRUE},
  {"version",     0, 0,             'v'},
  {"warn",        0, &warn_not_std, TRUE},

  {0, 0, 0, 0}
};


static void
usage (const char *progname)
{
  printf ("usage: %s [options] [file ...]\n%s%s%s%s%s%s%s%s%s", progname,
          "  -h  --help         print this usage and exit\n",
	  "  -e  --expression   eval commandline expression\n",
	  "  -c  --compile      output compiled dc script\n",
	  "  -i  --interactive  force interactive mode\n",
	  "  -l  --mathlib      use the predefined math routines\n",
	  "  -q  --quiet        don't print initial banner\n",
	  "  -s  --standard     non-standard bc constructs are errors\n",
	  "  -w  --warn         warn about non-standard bc constructs\n",
	  "  -v  --version      print version information and exit\n");
}


static void
parse_args (int argc, char **argv)
{
  int optch;
  int long_index;
  file_node *temp;
  int i;
  int size;

  /* Force getopt to initialize.  Depends on GNU getopt. */
  optind = 0;

  /* Parse the command line */
  while (1)
    {
      optch = getopt_long (argc, argv, "chilqswve", long_options, &long_index);

      if (optch == EOF)  /* End of arguments. */
	break;

      switch (optch)
	{
	case 0: /* Long option setting a var. */
	  break;

	case 'c':  /* compile only */
	  compile_only = TRUE;
	  break;

	case 'h':  /* help */
	  usage(argv[0]);
	  bc_exit (0);
	  break;

	case 'i':  /* force interactive */
	  interactive = TRUE;
	  break;

	case 'e':  /* commandline expression */
	  eval_expression = TRUE;
	  interactive = FALSE;
	  break;

	case 'l':  /* math lib */
	  use_math = TRUE;
	  break;

	case 'q':  /* quiet mode */
	  quiet = TRUE;
	  break;

	case 's':  /* Non standard features give errors. */
	  std_only = TRUE;
	  break;

	case 'v':  /* Print the version. */
	  show_bc_version ();
	  bc_exit (0);
	  break;

	case 'w':  /* Non standard features give warnings. */
	  warn_not_std = TRUE;
	  break;

	default:
	  usage(argv[0]);
	  bc_exit (1);
	}
    }

#ifdef QUIET
  quiet = TRUE;
#endif

  if (eval_expression) {
    size= argc+1;
    for (i=optind ; i<argc ; i++)
      size += strlen(argv[i]);
    cmdline_expression= bc_malloc(size);
    cmdline_expression[0]=0;
    for (i=optind ; i<argc ; i++) {
      if (i)
        strcat(cmdline_expression, " ");
      strcat(cmdline_expression, argv[i]);
    }
    strcat(cmdline_expression, "\n");

    return;
  }

  /* Add file names to a list of files to process. */
  while (optind < argc)
    {
      temp = bc_malloc(sizeof(file_node));
      temp->name = argv[optind];
      temp->next = NULL;
      if (last == NULL)
	file_names = temp;
      else
	last->next = temp;
      last = temp;
      optind++;
    }
}

/* The main program for bc. */
int
main (int argc, char **argv)
{
  char *env_value;
  char *env_argv[30];
  int   env_argc;
  
  /* Interactive? */
  if (isatty(0) && isatty(1)) 
    interactive = TRUE;

#ifdef HAVE_SETVBUF
  /* attempt to simplify interaction with applications such as emacs */
  (void) setvbuf(stdout, NULL, _IOLBF, 0);
#endif

  /* Environment arguments. */
  env_value = getenv ("BC_ENV_ARGS");
  if (env_value != NULL)
    {
      env_argc = 1;
      env_argv[0] = strdup("BC_ENV_ARGS");
      while (*env_value != 0)
	{
	  if (*env_value != ' ')
	    {
	      env_argv[env_argc++] = env_value;
	      while (*env_value != ' ' && *env_value != 0)
		env_value++;
	      if (*env_value != 0)
		{
		  *env_value = 0;
		  env_value++;
		}
	    }
	  else
	    env_value++;
	}
      parse_args (env_argc, env_argv);
    }

  /* Command line arguments. */
  parse_args (argc, argv);

  /* Other environment processing. */
  if (getenv ("POSIXLY_CORRECT") != NULL)
    std_only = TRUE;

  env_value = getenv ("BC_LINE_LENGTH");
  if (env_value != NULL)
    {
      line_size = atoi (env_value);
      if (line_size < 3 && line_size != 0)
	line_size = 70;
    }
  else
    line_size = 70;

  /* Initialize the machine.  */
  init_storage();
  init_load();

  /* Set up interrupts to print a message. */
  if (interactive)
    signal (SIGINT, use_quit);

  /* Initialize the front end. */
  init_tree();
  init_gen ();
  is_std_in = FALSE;
  first_file = TRUE;
  if (!open_new_file ())
    bc_exit (1);

#if defined(LIBEDIT)
  if (interactive) {
    /* Enable libedit support. */
    edit = el_init ("bc", stdin, stdout, stderr);
    hist = history_init();
    el_set (edit, EL_EDITOR, "emacs");
    el_set (edit, EL_HIST, history, hist);
    el_set (edit, EL_PROMPT, el_pmtfunc);
    el_source (edit, NULL);
    history (hist, &histev, H_SETSIZE, INT_MAX);
  }
#endif

#if defined(READLINE)
  if (interactive) {
    /* Readline support.  Set both application name and input file. */
    rl_readline_name = "bc";
    rl_instream = stdin;
    using_history ();
  }
#endif

  /* Do the parse. */
  yyparse ();

  /* End the compile only output with a newline. */
  if (compile_only)
    printf ("\n");

  bc_exit (0);
  return 0; // to keep the compiler from complaining
}


/* This is the function that opens all the files. 
   It returns TRUE if the file was opened, otherwise
   it returns FALSE. */

int
open_new_file (void)
{
  FILE *new_file;
  file_node *temp;

  /* Set the line number. */
  line_no = 1;

  /* Check to see if we are done. */
  if (is_std_in) return (FALSE);

  /* Open the other files. */
  if (use_math && first_file)
    {
      /* Load the code from a precompiled version of the math libarary. */
      CONST char **mstr;

      /* These MUST be in the order of first mention of each function.
	 That is why "a" comes before "c" even though "a" is defined after
	 after "c".  "a" is used in "s"! */
      (void) lookup (strdup("e"), FUNCT);
      (void) lookup (strdup("l"), FUNCT);
      (void) lookup (strdup("s"), FUNCT);
      (void) lookup (strdup("a"), FUNCT);
      (void) lookup (strdup("c"), FUNCT);
      (void) lookup (strdup("j"), FUNCT);
      mstr = libmath;
      while (*mstr) {
           load_code (*mstr);
	   mstr++;
      }
    }
  
  /* One of the argv values. */
  if (file_names != NULL)
    {
      new_file = fopen (file_names->name, "r");
      if (new_file != NULL)
	{
	  new_yy_file (new_file);
	  temp = file_names;
	  file_name  = temp->name;
	  file_names = temp->next;
	  free (temp);
	  return TRUE;
	}
      fprintf (stderr, "File %s is unavailable.\n", file_names->name);
      bc_exit (1);
    }
  
  /* If we fall through to here, we should return stdin. */
  new_yy_file (stdin);
  is_std_in = TRUE;
  return TRUE;
}


/* Set yyin to the new file. */

void
new_yy_file (FILE *file)
{
  if (!first_file) fclose (yyin);
  yyin = file;
  first_file = FALSE;
}


/* Message to use quit.  */

void
use_quit (int sig)
{
#ifdef DONTEXIT
  int save = errno;
  write (1, "\n(interrupt) use quit to exit.\n", 31);
  signal (SIGINT, use_quit);
  errno = save;
#else
  write (1, "\n(interrupt) Exiting bc.\n", 26);
  bc_exit(0);
#endif
}
